core/src/main/scala/filodb.core/memstore/ratelimit/CardinalityTracker.scala

```diff
@@ -102,14 +114,104 @@ class CardinalityTracker(ref: DatasetRef,
+   * 1. (total across all ws)
+   * 2. (my_ws)
+   * 3. (my_ws, my_ns)
+   * 4. (my_ws, my_ns, my_metric)
+   *
+   * @param shardKey
+   */
+  def updateCardinalityCountsDS(shardKey: Seq[String]): Unit = {
+    (0 to shardKey.length).foreach { i =>
+
+      // update current prefix's cardinality count
+      val prefix = shardKey.take(i)
+      val cardCountRecord = cardinalityCountMapDS.get(prefix)
+        .map(x => (x._1 + 1, x._2))
+        .getOrElse((1, 0)) // child prefix update parent's childrenCount
+      cardinalityCountMapDS.put(prefix, cardCountRecord)
+
+      // update children count of parent and throw exception if quota reached
+      if (i > 0) {
+        val parentPrefix = shardKey.take(i - 1)
+
+        // we always add parent before the child, hence it is okay to get the parent prefix's record directly
+        // without the None check
+        val updatedCountRecord = cardinalityCountMapDS.get(parentPrefix)
+          .map(x => (x._1, x._2 + 1)).get
+
+        // check if number of children is higher than the given quota. This allows us guard our physical resources
+        // and avoid failures because of runaway cardinality
+        val childrenQuota = defaultChildrenQuota(parentPrefix.length)
+        if (updatedCountRecord._2 > childrenQuota) {
+          quotaExceededProtocol.quotaExceeded(ref, shard, prefix, childrenQuota)
+          throw QuotaReachedException(prefix, prefix, childrenQuota)
+        }
+
+        // store the updated parent's childrenCount
+        cardinalityCountMapDS.put(parentPrefix, updatedCountRecord)
+      }
+    }
+
+    if (cardinalityCountMapDS.size > dsCardinalityMapFlushCount) {
+      flushCardinalityDataToRocksDB()
+    }
+  }
+
+  /*
+  * Flush the cardinality data to RocksDB before reading the counts. The downsample cardinality count is built from
+  * scratch at a periodic interval and the caller of CardinalityTracker can also call this method to ensure all data
+  * is flushed to RocksDB
+  * */
+  def flushCardinalityDataToRocksDB(): Unit = {
+    if (cardinalityCountMapDS.size > 0) {
+      // iterate through map and store each prefix and count to the rocksDB
+      cardinalityCountMapDS.foreach(kv => {
+        modifyCountDS(kv._1, kv._2._1, kv._2._1, kv._2._2)
+      })
+      // clear the map
+      cardinalityCountMapDS.clear()
+    }
+  }
+
+  /**
+   * Used to store the cardinality count for the given prefix in the downsample cluster.
+   * NOTE: In downsample cluster, tsCount == activeTsCount. So totalDelta and activeDelta is same
+   * @param prefix usually contains labels _ws_, _ns_, _metric_ and different combinations of it
+   * @param totalDelta Increase in total timeseries
+   * @param activeDelta Increase in active timeseries
+   * @param childrenDelta Increase in children count
+   */
+  private def modifyCountDS(prefix: Seq[String],
+                            totalDelta: Int, activeDelta: Int, childrenDelta: Int): Unit = synchronized {
```

Comment:
rocksdb should be thread safe.
I think the race here is that different threads may updating the same key.
The lock granularity could be key level.

Reply:
We are not planning on calling this from multiple threads as of now, but just to be safe we added a function scope synchronized. Can be optimized, when we plan to call this from multiple threads.

Reply:
Just to be safe, can we add `synchronized` to all methods that directly read from / write to the `Map`? (Especially since we don't expect contention to be an issue-- we will future-proof ourselves from race conditions with very little cost).

Reply:
Makes sense. let me do this